app/toolbox-server.hs

```diff
@@ -0,0 +1,93 @@
+module Main (main) where
+
+import Concur.Core
+  ( Widget,
+    liftSTM,
+  )
+import Concur.Replica
+  ( div,
+    pre,
+    runDefault,
+    text,
+  )
+import Control.Applicative ((<|>))
+import Control.Concurrent (forkIO)
+import Control.Concurrent.STM
+  ( TVar,
+    atomically,
+    newTVar,
+    readTVar,
+    retry,
+    writeTVar,
+  )
+import Data.Map (Map)
+import qualified Data.Map as M
+import Data.Text (Text)
+import qualified Data.Text as T
+import Replica.VDOM.Types (HTML)
+import Toolbox.Comm
+  ( receiveObject,
+    runServer,
+  )
+import Prelude hiding (div)
+
+main :: IO ()
+main = do
+  var <- atomically $ newTVar Nothing
+  _ <- forkIO $ listener var
+  webServer var
+
+type ModuleMessages = Map Text Text
+
+listener :: TVar (Maybe (Int, ModuleMessages)) -> IO ()
+listener var =
+  runServer
+    "/tmp/ghc-build-analyzer.ipc"
+    ( \sock -> do
+        (modName, msg) <- receiveObject sock
+        -- threadDelay 500_000
+        updateModuleMessages var (modName, msg)
+    )
+
+updateModuleMessages :: TVar (Maybe (Int, ModuleMessages)) -> (Text, Text) -> IO ()
+updateModuleMessages var (modName, msg) = do
+  mmap <- atomically $ readTVar var
+  case mmap of
+    Nothing -> do
+      let m = M.singleton modName msg
+      atomically $ writeTVar var (Just (1, m))
```

Comment:
It seems like `Nothing` is semantically equivalent to `(0, mempty)`. I'm curious if that's a viable option too - may simplify the access patterns here

Reply:
Changed as suggested :-)